### PR TITLE
SI-9143 Avoid edge case in BigDecimal factory

### PR DESCRIPTION
**Pull request by som-snytt**
The Java factory fails on MinValue exponent, as defined,

```
"0E+2147483648"
```

The JDK source says

```
// Next test is required for backwards compatibility
```

which suggests that Scala could handle it gracefully.

Where grace == I don't give a flying backwards compatibility.

The issue was whether ScalaCheck is free to generate values
that can't be reconstituted via the string factory.

http://stackoverflow.com/questions/28461403/scala-bigdecimal-fails-to-translate-a-decimal-string-representation-created-by-i/28461644

Review by @Ichoran

### CODE REVIEW

File: src/library/scala/math/BigDecimal.scala [hunk 1]
```diff
@@ -122,7 +122,18 @@ object BigDecimal {
...
    *  specified in a `String`.
    */
-  def exact(s: String): BigDecimal = exact(new BigDec(s))
+  def exact(s: String): BigDecimal =
+    try exact(new BigDec(s))
+    catch {
+      // if it fails because exp is MinValue, construct by hand.
+      // BigDec throws for reasons of backward compatibility.
+      case e: NumberFormatException if s endsWith "+2147483648" =>
+        val r = """(\+|-)?(\d+)(\.\d+)?(e|E)\+2147483648""".r
```
Comment (Ichoran) [inline src/library/scala/math/BigDecimal.scala:131]: The `+` might be expected, but it's not actually necessary for the BigDecimal constructor, so you should just take all the NumberFormatExceptions and match on r (and make `+` optional).  This has cleaner logic also: you don't have to worry about long strings of numbers in two spots.

**Comment by som-snytt** [comment]
@Ichoran Would you like me to fix this up? I had second thoughts about the use case.

**Comment by Ichoran** [comment]
@som-snytt That would be great.  It's not exceedingly important, of course, but it's more satisfying to not have edge cases lurking around than to get cut on the edges now and then.
